x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java

```diff
@@ -45,6 +47,9 @@ protected TypeResolution resolveType() {
             if (DataTypeConversion.commonType(l, r) == null) {
                 return new TypeResolution(format("[{}] has arguments with incompatible types [{}] and [{}]", symbol(), l, r));
             } else {
+                if (function() == SUB && right().dataType().isDateBased() && DataTypes.isInterval(left().dataType())) {
```

Comment:
Here you can do `if (function() == SUB && r.isDateBased() && DataTypes.isInterval(l))` (`r` and `l` are already defined above).

Reply:
The type check should happen inside SUB not in its parent class...